lib/rules/no-lone-blocks.js

```diff
@@ -11,17 +13,94 @@
 
 module.exports = function(context) {
 
-    return {
-        "BlockStatement": function (node) {
-            // Check for any occurrence of BlockStatement > BlockStatement or
-            // Program > BlockStatement
-            var parent = context.getAncestors().pop();
-            if (parent.type === "BlockStatement" || parent.type === "Program") {
-                context.report(node, "Block is nested inside another block.");
+    // The lone blocks to be checked for block-level bindings
+    var loneBlocks = [],
+        ruleDef;
+
+    /**
+     * Reports a node as invalid.
+     * @param {ASTNode} node - The node to be reported.
+     * @returns {void}
+    */
+    function report(node) {
+        context.report(node, "Block is nested inside another block.");
+    }
+
+    /**
+     * Checks for any ocurrence of BlockStatement > BlockStatement or Program > BlockStatement
+     * @returns {boolean} True if the current node is a lone block.
+    */
+    function isLoneBlock() {
+        var parent = context.getAncestors().pop();
+        return parent.type === "BlockStatement" || parent.type === "Program";
+    }
+
+    /**
+     * Checks the enclosing block of the current node for block-level bindings,
+     * and "marks it" as valid if any.
+     * @returns {void}
+    */
+    function markLoneBlock() {
+        if (loneBlocks.length === 0) {
+            return;
+        }
+
+        var block = context.getAncestors().pop(),
+            index;
+
+        if (!isLoneBlock(block)) {
+            return;
+        }
+
+        index = loneBlocks.indexOf(block);
+
+        if (index >= 0) {
+            loneBlocks.splice(index, 1);
+        }
+    }
+
+    // Default rule definition: report all lone blocks
+    ruleDef = {
+        BlockStatement: function(node) {
+            if (isLoneBlock(node)) {
+                report(node);
             }
         }
     };
 
+    // ES6: report blocks without block-level bindings
+    if (context.ecmaFeatures.blockBindings || context.ecmaFeatures.classes) {
+        ruleDef = {
+            "BlockStatement": function(node) {
+                if (isLoneBlock(node)) {
+                    loneBlocks.push(node);
+                }
+            },
+            "Program:exit": function() {
```

Comment:
Hum, I see now that I could save some memory by reporting a block at `"BlockStatement:exit"` and removing it from `loneBlocks`. Something like:

```js
"BlockStatement:exit": function(node) {
    if (!isLoneBlock(node)) {
        return;
    }

    var index = loneBlocks.indexOf(node);

    if (index >= 0) {
        report(node);
        loneBlocks.splice(index, 1);
    }
}
```

However, I don't know how we could remove the `loneBlocks` list completely. Is the AST traversal order specified?